### PR TITLE
Faster display list transfer

### PR DESCRIPTION
**Pull request by pcwalton**
This series of commits improves performance of display list construction in browser.html by about 3x when using WebRender.

It requires https://github.com/servo/webrender_traits/pull/18 and https://github.com/servo/webrender/pull/231.

Anyone should feel free to review if they have time; I'll ask someone in particular once those two upstream commits land.

cc @paulrouget

---

This change is [Review on Reviewable](https://reviewable.io/reviews/servo/servo/9947)

### CODE REVIEW

**Comment by Ms2ger** [comment]
```
/home/travis/build/servo/servo/components/layout/layout_thread.rs:958:25: 958:50 error: this function takes 6 parameters but 8 parameters were supplied [E0061]

/home/travis/build/servo/servo/components/layout/layout_thread.rs:958                     api.set_root_stacking_context(sc_id,

                                                                                              ^~~~~~~~~~~~~~~~~~~~~~~~~

/home/travis/build/servo/servo/components/layout/layout_thread.rs:958:25: 958:50 help: run `rustc --explain E0061` to see a detailed explanation

/home/travis/build/servo/servo/components/layout/webrender_helpers.rs:561:22: 561:46 error: no method named `next_stacking_context_id` found for type `&mut webrender_traits::api::RenderApi` in the current scope

/home/travis/build/servo/servo/components/layout/webrender_helpers.rs:561         let id = api.next_stacking_context_id();

                                                                                               ^~~~~~~~~~~~~~~~~~~~~~~~

/home/travis/build/servo/servo/components/layout/webrender_helpers.rs:571:22: 571:42 error: no method named `next_display_list_id` found for type `&mut webrender_traits::api::RenderApi` in the current scope

/home/travis/build/servo/servo/components/layout/webrender_helpers.rs:571         let id = api.next_display_list_id();

                                                                                               ^~~~~~~~~~~~~~~~~~~~

/home/travis/build/servo/servo/components/layout/webrender_helpers.rs:573:63: 573:73 error: no method named `descriptor` found for type `webrender_traits::display_list::BuiltDisplayList` in the current scope

/home/travis/build/servo/servo/components/layout/webrender_helpers.rs:573                                                  display_list.descriptor().has_stacking_contexts;

                                                                                                                                        ^~~~~~~~~~

error: aborting due to 4 previous errors

Could not compile `layout`.
```

**Comment by jdm** [comment]
The webrender-related PRs have merged successfully.

**Comment by pcwalton** [comment]
Updated to include the WR updates.

r? @glennw

**Comment by pcwalton** [comment]
Fixed the tests by unbreaking the synchronous image loading path.

r? @glennw

**Comment by jdm** [comment]
```
  ▶ TIMEOUT [expected PASS] /_mozilla/css/filter_inline_a.html
  │ 
  │ thread 'LayoutWorker worker 1/6' panicked at 'called `Result::unwrap()` on an `Err` value: IoError(Error { repr: Custom(Custom { kind: Other, error: StringError("Unknown Mach error: 46") }) })', ../src/libcore/result.rs:746
  │ stack backtrace:
  │    1:        0x10abae798 - sys::backtrace::tracing::imp::write::h3153bab3099b48114Zu
  │    2:        0x10abb2389 - panicking::default_handler::_$u7b$$u7b$closure$u7d$$u7d$::closure.43920
  │    3:        0x10abb1fb8 - panicking::default_handler::h1183b2a11c8682560zz
  │    4:        0x10ab9d386 - sys_common::unwind::begin_unwind_inner::h1f97bc9e4cfe0396LZt
  │    5:        0x10ab9e3be - sys_common::unwind::begin_unwind_fmt::hda8646f04aa67d8aRYt
  │    6:        0x10abadb37 - rust_begin_unwind
  │    7:        0x10abd7f90 - panicking::panic_fmt::h40f5ec0cdc3fc429FRL
  │    8:        0x1093c2507 - result::unwrap_failed::h1240322561869761750
  │    9:        0x1093c9b3b - context::LayoutContext<'a>::get_or_request_image_or_meta::h2374b6453ff84c53FJe
  │   10:        0x109364977 - fragment::ImageFragmentInfo::new::h8384581981763288016
  │   11:        0x109363e8e - construct::FlowConstructor<'a, ConcreteThreadSafeLayoutNode>::build_fragment_for_block::h16212892555337533501
  │   12:        0x109358809 - construct::FlowConstructor<'a, ConcreteThreadSafeLayoutNode>.PostorderNodeMutTraversal<ConcreteThreadSafeLayoutNode>::process::h10160195791494672944
  │   13:        0x109353554 - traversal::RecalcStyleAndConstructFlows<'lc>.DomTraversalContext<ServoLayoutNode<'ln>>::process_postorder::hac3607a05e3a306ffKx
  │   14:        0x1094a497f - parallel::top_down_dom::h5484319701395010818
  │   15:        0x1093fdbad - sys_common::unwind::try::try_fn::h336964700432214359
  │   16:        0x10abadacb - __rust_try
  │   17:        0x10abada53 - sys_common::unwind::inner_try::hdf6affff6fc30ecdNWt
  │   18:        0x1093fee07 - boxed::F.FnBox<A>::call_box::h8246899095210445003
  │   19:        0x10abb146c - sys::thread::Thread::new::thread_start::he33a0e5f2597fcdejOy
  │   20:     0x7fff86bcf059 - _pthread_body
  └   21:     0x7fff86bcefd6 - _pthread_start

  ▶ TIMEOUT [expected PASS] /_mozilla/css/inline_margin_multiple_fragments_a.html
  │ 
  │ thread 'LayoutWorker worker 1/6' panicked at 'called `Result::unwrap()` on an `Err` value: IoError(Error { repr: Custom(Custom { kind: Other, error: StringError("Unknown Mach error: 46") }) })', ../src/libcore/result.rs:746
  │ stack backtrace:
  │    1:        0x110f6a798 - sys::backtrace::tracing::imp::write::h3153bab3099b48114Zu
  │    2:        0x110f6e389 - panicking::default_handler::_$u7b$$u7b$closure$u7d$$u7d$::closure.43920
  │    3:        0x110f6dfb8 - panicking::default_handler::h1183b2a11c8682560zz
  │    4:        0x110f59386 - sys_common::unwind::begin_unwind_inner::h1f97bc9e4cfe0396LZt
  │    5:        0x110f5a3be - sys_common::unwind::begin_unwind_fmt::hda8646f04aa67d8aRYt
  │    6:        0x110f69b37 - rust_begin_unwind
  │    7:        0x110f93f90 - panicking::panic_fmt::h40f5ec0cdc3fc429FRL
  │    8:        0x10f77e507 - result::unwrap_failed::h1240322561869761750
  │    9:        0x10f785b3b - context::LayoutContext<'a>::get_or_request_image_or_meta::h2374b6453ff84c53FJe
  │   10:        0x10f720977 - fragment::ImageFragmentInfo::new::h8384581981763288016
  │   11:        0x10f71fe8e - construct::FlowConstructor<'a, ConcreteThreadSafeLayoutNode>::build_fragment_for_block::h16212892555337533501
  │   12:        0x10f714809 - construct::FlowConstructor<'a, ConcreteThreadSafeLayoutNode>.PostorderNodeMutTraversal<ConcreteThreadSafeLayoutNode>::process::h10160195791494672944
  │   13:        0x10f70f554 - traversal::RecalcStyleAndConstructFlows<'lc>.DomTraversalContext<ServoLayoutNode<'ln>>::process_postorder::hac3607a05e3a306ffKx
  │   14:        0x10f86097f - parallel::top_down_dom::h5484319701395010818
  │   15:        0x10f7b9bad - sys_common::unwind::try::try_fn::h336964700432214359
  │   16:        0x110f69acb - __rust_try
  │   17:        0x110f69a53 - sys_common::unwind::inner_try::hdf6affff6fc30ecdNWt
  │   18:        0x10f7bae07 - boxed::F.FnBox<A>::call_box::h8246899095210445003
  │   19:        0x110f6d46c - sys::thread::Thread::new::thread_start::he33a0e5f2597fcdejOy
  │   20:     0x7fff86bcf059 - _pthread_body
  └   21:     0x7fff86bcefd6 - _pthread_start

  ▶ TIMEOUT [expected PASS] /_mozilla/css/no-image.html
  │ 
  │ thread 'LayoutWorker worker 1/6' panicked at 'called `Result::unwrap()` on an `Err` value: IoError(Error { repr: Custom(Custom { kind: Other, error: StringError("Unknown Mach error: 46") }) })', ../src/libcore/result.rs:746
  │ stack backtrace:
  │    1:        0x105d36798 - sys::backtrace::tracing::imp::write::h3153bab3099b48114Zu
  │    2:        0x105d3a389 - panicking::default_handler::_$u7b$$u7b$closure$u7d$$u7d$::closure.43920
  │    3:        0x105d39fb8 - panicking::default_handler::h1183b2a11c8682560zz
  │    4:        0x105d25386 - sys_common::unwind::begin_unwind_inner::h1f97bc9e4cfe0396LZt
  │    5:        0x105d263be - sys_common::unwind::begin_unwind_fmt::hda8646f04aa67d8aRYt
  │    6:        0x105d35b37 - rust_begin_unwind
  │    7:        0x105d5ff90 - panicking::panic_fmt::h40f5ec0cdc3fc429FRL
  │    8:        0x10454a507 - result::unwrap_failed::h1240322561869761750
  │    9:        0x104551b3b - context::LayoutContext<'a>::get_or_request_image_or_meta::h2374b6453ff84c53FJe
  │   10:        0x1044ec977 - fragment::ImageFragmentInfo::new::h8384581981763288016
  │   11:        0x1044ebe8e - construct::FlowConstructor<'a, ConcreteThreadSafeLayoutNode>::build_fragment_for_block::h16212892555337533501
  │   12:        0x1044e0809 - construct::FlowConstructor<'a, ConcreteThreadSafeLayoutNode>.PostorderNodeMutTraversal<ConcreteThreadSafeLayoutNode>::process::h10160195791494672944
  │   13:        0x1044db554 - traversal::RecalcStyleAndConstructFlows<'lc>.DomTraversalContext<ServoLayoutNode<'ln>>::process_postorder::hac3607a05e3a306ffKx
  │   14:        0x10462c97f - parallel::top_down_dom::h5484319701395010818
  │   15:        0x104585bad - sys_common::unwind::try::try_fn::h336964700432214359
  │   16:        0x105d35acb - __rust_try
  │   17:        0x105d35a53 - sys_common::unwind::inner_try::hdf6affff6fc30ecdNWt
  │   18:        0x104586e07 - boxed::F.FnBox<A>::call_box::h8246899095210445003
  │   19:        0x105d3946c - sys::thread::Thread::new::thread_start::he33a0e5f2597fcdejOy
  │   20:     0x7fff86bcf059 - _pthread_body
  └   21:     0x7fff86bcefd6 - _pthread_start

  ▶ TIMEOUT [expected PASS] /_mozilla/css/no_image_background_a.html
  │ 
  │ thread 'LayoutThread PipelineId { namespace_id: PipelineNamespaceId(0), index: PipelineIndex(0) }' panicked at 'called `Result::unwrap()` on an `Err` value: IoError(Error { repr: Custom(Custom { kind: Other, error: StringError("Unknown Mach error: 46") }) })', ../src/libcore/result.rs:746
  │ stack backtrace:
  │    1:        0x110b53798 - sys::backtrace::tracing::imp::write::h3153bab3099b48114Zu
  │    2:        0x110b57389 - panicking::default_handler::_$u7b$$u7b$closure$u7d$$u7d$::closure.43920
  │    3:        0x110b56fb8 - panicking::default_handler::h1183b2a11c8682560zz
  │    4:        0x110b42386 - sys_common::unwind::begin_unwind_inner::h1f97bc9e4cfe0396LZt
  │    5:        0x110b433be - sys_common::unwind::begin_unwind_fmt::hda8646f04aa67d8aRYt
  │    6:        0x110b52b37 - rust_begin_unwind
  │    7:        0x110b7cf90 - panicking::panic_fmt::h40f5ec0cdc3fc429FRL
  │    8:        0x10f367507 - result::unwrap_failed::h1240322561869761750
  │    9:        0x10f36eb3b - context::LayoutContext<'a>::get_or_request_image_or_meta::h2374b6453ff84c53FJe
  │   10:        0x10f36f118 - context::LayoutContext<'a>::get_webrender_image_for_url::hc15a9333a454ef091Ke
  │   11:        0x10f31b229 - display_list_builder::Fragment.FragmentDisplayListBuilding::build_display_list_for_background_image::h9cdcbfc148f9e07f0mf
  │   12:        0x10f31a523 - display_list_builder::Fragment.FragmentDisplayListBuilding::build_display_list_for_background_if_applicable::h727b8327dff2e9d1Gff
  │   13:        0x10f31db12 - display_list_builder::Fragment.FragmentDisplayListBuilding::build_display_list::h97e5d462cd7fc90c6Of
  │   14:        0x10f319431 - display_list_builder::BlockFlow.BlockFlowDisplayListBuilding::build_display_list_for_block::h297d266e39f1e95e1rg
  │   15:        0x10f3171fa - block::BlockFlow.Flow::build_display_list::h400ab1c5bf772a43zqc
  │   16:        0x10f30cda1 - traversal::BuildDisplayList<'a>::traverse::h17468936ac4ca059UWx
  │   17:        0x10f30cdf7 - traversal::BuildDisplayList<'a>::traverse::h17468936ac4ca059UWx
  │   18:        0x10f30cdf7 - traversal::BuildDisplayList<'a>::traverse::h17468936ac4ca059UWx
  │   19:        0x10f354a38 - sequential::build_display_list_for_subtree::h7c60906b13e3ae81CDt
  │   20:        0x10f44c41b - layout_thread::LayoutThread::perform_post_main_layout_passes::h4c543d7e03895829XDr
  │   21:        0x10f3ea756 - layout_thread::LayoutThread::perform_post_style_recalc_layout_passes::hb317b3c6ee2a3a85zAr
  │   22:        0x10f3f0e42 - layout_thread::LayoutThread::handle_reflow::h69da426ca17c1ae5Ver
  │   23:        0x10f3d1255 - layout_thread::LayoutThread::handle_request_helper::hed0e7f425cf79f04JRq
  │   24:        0x10f3b674b - layout_thread::LayoutThread::handle_request::h158f01118fc11f158Lq
  │   25:        0x10f37cf4a - sys_common::unwind::try::try_fn::h2571472373718225
  │   26:        0x110b52acb - __rust_try
  │   27:        0x110b52a53 - sys_common::unwind::inner_try::hdf6affff6fc30ecdNWt
  │   28:        0x10f37d8a1 - boxed::F.FnBox<A>::call_box::h2680565343670644496
  │   29:        0x110b5646c - sys::thread::Thread::new::thread_start::he33a0e5f2597fcdejOy
  │   30:     0x7fff86bcf059 - _pthread_body
  │   31:     0x7fff86bcefd6 - _pthread_start
```

File: components/layout/context.rs [hunk 1]
```diff
@@ -132,45 +138,35 @@ impl<'a> LayoutContext<'a> {
...
 
-    pub fn get_or_request_image(&self, url: Url, use_placeholder: UsePlaceholder)
-                                -> Option<Arc<Image>> {
+    fn get_or_request_image_synchronously(&self, url: Url) -> Option<Arc<Image>> {
+        debug_assert!(opts::get().output_file.is_some() || opts::get().exit_after_load);
+
         // See if the image is already available
-        let result = self.shared.image_cache_thread.find_image(url.clone(),
-                                                             use_placeholder);
+        let result = self.shared.image_cache_thread.find_image(url.clone(), UsePlaceholder::No);
```
Comment (jdm) [inline components/layout/context.rs]: Why the switch to UsePlaceholder::No here?

**Comment by pcwalton** [comment]
@jdm @glennw Added `use_placeholder` back. (I don't think it affects anything either way, but anyway.)

r? @glennw

**Comment by pcwalton** [comment]
cc @larsbergstrom

**Comment by larsbergstrom** [comment]
@bors-servo retry
- infra

**Pull request merged**